app/models/assignment.rb

```diff
@@ -1464,4 +1464,68 @@ def same_membership_as_csv_row?(row, existing_grouping)
     end
   end
 
+  # Reformatted logic for download_assignment_list controller method
+  # Returns an output file for controller to handle.
+  def self.get_assignment_list(file_format, default_fields)
+    assignments = self.all
+
+    case file_format
+    when 'yml'
+      map = {}
+      map[:assignments] = []
+      assignments.map do |assignment|
+        m = {}
+        default_fields.length.times do |i|
+          m[default_fields[i]] = assignment.send(default_fields[i])
+        end
+        map[:assignments] << m
+      end
+      output = map.to_yaml
+    when 'csv'
+      output = MarkusCSV.generate(assignments) do |assignment|
+        default_fields.map do |f|
+          assignment.send(f.to_s)
+        end
+      end
+    end
+    return output
+  end
+
+  def self.upload_assignment_list(file_format, assignment_list, default_fields)
+
+  case file_format
+  when 'csv'
+    result = MarkusCSV.parse(assignment_list) do |row|
+    assignment = self.find_or_create_by(short_identifier: row[0])
+    attrs = Hash[default_fields.zip(row)]
+    attrs.delete_if { |_, v| v.nil? }
+      if assignment.new_record?
+         assignment.submission_rule = NoLateSubmissionRule.new
+         assignment.assignment_stat = AssignmentStat.new
+         assignment.token_period = 1
+         assignment.non_regenerating_tokens = false
```

Comment:
These may be default values. Check with @hl197 to see if they can be removed.

Reply:
```suggestion
         assignment.non_regenerating_tokens = false
```
can be removed since it is the default.